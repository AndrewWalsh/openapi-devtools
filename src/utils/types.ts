import type { Schema } from "genson-js";
import { SecuritySchemeType } from "openapi3-ts/oas31";
import type { RadixRouter } from "radix3";

export type JSONType =
  | string
  | number
  | boolean
  | null
  | { [key: string]: JSONType }
  | Array<JSONType>;

// Unique identifier for each type of authentication, aligned to oai31
export enum AuthType {
  HTTP_HEADER_BEARER = "Bearer",
  HTTP_HEADER_BASIC = "Basic",
  HTTP_HEADER_DIGEST = "Digest",
  APIKEY_HEADER_ = "APIKEY_HEADER_",
  APIKEY_COOKIE_ = "APIKEY_COOKIE_",
}

// Can be dynamic e.g. APIKEY_HEADER_ + name
export type AuthTypeString = AuthType | string;

// Modelled on a Security Scheme Object https://spec.openapis.org/oas/v3.1.0#security-scheme-object
export interface Authentication {
  // So there is a straightforward way of identifying the type
  authType: AuthTypeString;
  // One of: 'apiKey' | 'http' | 'oauth2' | 'openIdConnect'
  type: SecuritySchemeType;
  // Could potentially generate some form of description from token values
  description?: string;
  // The name of the header, query or cookie parameter to be used. Used only with apiKey
  name?: string;
  // The location of the API key, "query", "header" or "cookie"
  in: "query" | "header" | "cookie";
  // A valid HTTP auth scheme defined in RFC9110. Only used when type = http
  scheme?: "Basic" | "Bearer" | "Digest";
}

// A Leaf stores data about an endpoint
export type Leaf = {
  // Name is based on authType, which should be unique per authentication variant
  authentication?: {
    [name: AuthTypeString]: Authentication;
  };
  // The current pathname of this endpoint, which may be static or parameterised
  pathname: string;
  // Methods such as GET, POST and schema values for requests
  methods: {
    [method: string]: {
      [statusCode: string]: {
        // Requests may not contain a body
        request?: {
          // mediaType is a a mime type such as application/json
          [mediaType: string]: {
            body?: Schema;
            // Sample of the most recent request
            mostRecent?: unknown;
          };
        };
        requestHeaders?: Schema;
        response: {
          [mediaType: string]: {
            body?: Schema;
            // Sample of the most recent response
            mostRecent?: unknown;
          };
        };
        responseHeaders?: Schema;
        queryParameters?: Schema;
      };
    };
  };
};

export enum PartType {
  Static,
  Dynamic,
}
export type Parts = Array<{ part: string; type: PartType }>;
// An Endpoint wraps a Leaf with additional data. Used as an intermediary step to simplify conversion into OAI.
export type Endpoint = {
  // A host e.g. example.com
  host: string;
  // The full path including parameters, such as /1/:param1/:param2/4
  pathname: string;
  // An array of parts such as 1, 2, and 3 in /1/2/3. Includes their type, static or dynamic (parameter)
  parts: Parts;
  // Data for this endpoint
  data: Leaf;
};

export enum Status {
  INIT,
  RECORDING,
  STOPPED,
}

export type RouteData = { data: Leaf };
type Router = RadixRouter<RouteData>;
export type RouterMap = { [host: string]: Router };
export type LeafMap = { [host: string]: Record<string, RouteData> };

export type EndpointsByHost = Array<{ endpoints: Endpoint[]; host: string }>;
