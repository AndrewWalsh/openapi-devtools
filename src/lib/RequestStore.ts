import { JSONType, RouterMap, LeafMap, Endpoint } from "../utils/types";
import { parameterise, insertLeafMap, upsert } from "./store-helpers";
import { omit, unset } from "lodash";
import leafMapToEndpoints from "./leafmap-to-endpoints";
import stringify from "json-stable-stringify";

/**
 * RequestStore handles routing to endpoints
 * Optimised for fast lookups & insertion via a Radix Tree
 */
export default class RequestStore {
  private store: RouterMap;
  private leafMap: LeafMap;
  private disabledHosts: Set<string>;

  constructor() {
    this.leafMap = {}; // persist.get() || {};
    this.store = {}; // leafMapToRouterMap(this.leafMap);
    this.disabledHosts = new Set();
  }

  public import(json: string): boolean {
    try {
      const { store, leafMap, disabledHosts } = JSON.parse(json);
      this.disabledHosts = new Set(disabledHosts);
      this.store = store;
      this.leafMap = leafMap;
      return true;
    } catch {
      return false;
    }
  }

  public export = (): string => {
    return stringify({
      store: this.store,
      leafMap: this.leafMap,
      disabledHosts: Array.from(this.disabledHosts),
    });
  };

  public clear(): void {
    this.store = {};
    this.leafMap = {};
    this.disabledHosts = new Set();
    // persist.clear();
  }

  public endpoints(): Array<Endpoint> {
    const withoutDisabled = omit(
      this.leafMap,
      Array.from(this.disabledHosts)
    ) as Readonly<typeof this.leafMap>;
    return leafMapToEndpoints(withoutDisabled);
  }

  public get(): Readonly<RouterMap> {
    return omit(this.store, Array.from(this.disabledHosts)) as Readonly<
      typeof this.store
    >;
  }

  public hosts(): Array<string> {
    return Object.keys(this.store);
  }

  public insert(
    harRequest: chrome.devtools.network.Request,
    responseBody: JSONType
  ) {
    const result = upsert({ harRequest, responseBody, store: this.store });
    if (!result) return;
    const { insertedPath, insertedLeaf, insertedHost } = result;
    insertLeafMap({
      leafMap: this.leafMap,
      host: insertedHost,
      leaf: insertedLeaf,
      path: insertedPath,
    });
    // persist.set(this.leafMap);
  }

  public parameterise(index: number, path: string, host: string): void {
    const result = parameterise({ store: this.store, index, path, host });
    if (!result) return;
    const { removedPaths, insertedPath, insertedLeaf } = result;
    const unsetLeafMap = (path: string) => unset(this.leafMap[host], path);
    removedPaths.concat([path]).forEach(unsetLeafMap);
    insertLeafMap({
      leafMap: this.leafMap,
      host,
      leaf: insertedLeaf,
      path: insertedPath,
    });
    // persist.set(this.leafMap);
  }

  public setDisabledHosts(disabledHosts: Set<string>): void {
    this.disabledHosts = disabledHosts;
  }
}
