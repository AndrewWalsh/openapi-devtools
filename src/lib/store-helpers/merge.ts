import { mergeSchemas, Schema } from "genson-js";

import type { Leaf } from "../../utils/types";

type Data = Leaf["methods"]["get"];
type Req = NonNullable<Leaf["methods"]["get"]["request"]>;
type Res = Leaf["methods"]["get"]["response"];

const mergeAuthentication = (dest: Leaf, src: Leaf): void => {
  if (!src.authentication) return;
  if (!dest.authentication) dest.authentication = {};
  Object.entries(src.authentication).forEach(([key, value]) => {
    dest.authentication![key] = value;
  });
};

const mergeRequest = (dest: Data, src: Req = {}) => {
  Object.entries(src).forEach(([mediaType, srcData]) => {
    // Nothing in dest or src to merge
    if (!srcData || !srcData.body) return;
    // Nothing in dest to merge, but src has data
    if (dest["request"]?.[mediaType]) {
      dest["request"][mediaType] = srcData;
    } else {
      // Both src and dest have data
      dest["request"]![mediaType]!.body = mergeSchemas([
        dest["request"]![mediaType]!.body!,
        srcData.body,
      ]);
      dest["request"]![mediaType]!.mostRecent = srcData.mostRecent;
    }
  });
};

const mergeResponse = (dest: Data, src: Res = {}) => {
  Object.entries(src).forEach(([statusCode, srcMediaTypeObj]) => {
    // statusCode in src does not exist in dest
    if (!dest["response"]?.[statusCode]) {
      dest["response"][statusCode] = srcMediaTypeObj;
      return;
    }
    // statusCode exists in both dest and src
    Object.entries(srcMediaTypeObj).forEach(([mediaType, mediaTypeData]) => {
      const srcData = srcMediaTypeObj[mediaType]!;
      if (!dest["response"]?.[statusCode]?.[mediaType]) {
        // dest does not contain mediaType, set from src
        dest["response"][statusCode]![mediaType] = srcData;
        return;
      } else {
        // merge schemas
        dest["response"][statusCode]![mediaType]!.body = mergeSchemas([
          dest["response"]![statusCode]![mediaType]!.body!,
          srcData.body || {},
        ]);
        dest["response"]![statusCode]![mediaType]!.mostRecent = mediaTypeData.mostRecent;
      }
    });
  });
};

export const mergeLeaves = (dest: Leaf, src: Leaf): Leaf => {
  mergeAuthentication(dest, src);
  for (const [method, methodObj] of Object.entries(src.methods)) {
    // Method doesn't exist in dest, set src and continue to next method
    if (!dest.methods[method]) {
      dest.methods[method] = methodObj;
      continue;
    }
    const srcSchema = src.methods[method]!;
    const destSchema = dest.methods[method]!;
    // Merge request
    if (destSchema.request || srcSchema.request) {
      mergeRequest(destSchema, srcSchema.request);
    }

    // Merge query params
    if (destSchema.queryParameters || srcSchema.queryParameters) {
      const schemas = [
        destSchema.queryParameters,
        srcSchema.queryParameters,
      ].filter(Boolean) as Schema[];
      destSchema.queryParameters = mergeSchemas(schemas);
    }

    // Merge request headers
    if (destSchema.requestHeaders || srcSchema.requestHeaders) {
      const schemas = [
        destSchema.requestHeaders,
        srcSchema.requestHeaders,
      ].filter(Boolean) as Schema[];
      destSchema.requestHeaders = mergeSchemas(schemas);
    }

    // Merge response headers
    if (destSchema.responseHeaders || srcSchema.responseHeaders) {
      const schemas = [
        destSchema.responseHeaders,
        srcSchema.responseHeaders,
      ].filter(Boolean) as Schema[];
      destSchema.responseHeaders = mergeSchemas(schemas);
    }

    // Merge responses
    mergeResponse(destSchema, methodObj.response);
  }
  return dest;
};
