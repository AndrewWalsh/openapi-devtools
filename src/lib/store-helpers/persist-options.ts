import store2 from "store2";
import type { Options } from "../RequestStore";

const namespace = "openapi-devtools-options";
const key = "options";
const store = store2.namespace(namespace);

export const defaultOptions: Readonly<Options> = {
  enableMoreInfo: false,
};

const get = (): Options => {
  const options = store.get(key);
  if (options) return options;
  store.set(key, defaultOptions);
  return defaultOptions;
};
const set = (value: Partial<Options>) => {
  const options = get();
  const newOptions = { ...options, ...value };
  store.set(key, newOptions);
};
const clear = () => store.clear();

export default {
  get,
  set,
  clear,
};
