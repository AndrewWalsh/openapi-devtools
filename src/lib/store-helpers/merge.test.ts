import { cloneDeep } from "lodash";
import { expect, it } from "vitest";

import postJson from "../__fixtures__/post-application-json.js";
import { Options } from "../RequestStore.js";
import { JSONType } from "../../utils/types.js";
import createLeaf from "./create-leaf.js";
import { mergeLeaves } from "./merge.js";
import { defaultOptions } from "./persist-options.js";

const setup = (options: Options) => {
  const harRequest1 = postJson;
  const harRequest2 = cloneDeep(postJson);
  harRequest2.request.postData!.text = '{ "test": true }';
  const responseBody: JSONType = { test: 1 };
  const leaf1 = createLeaf({ harRequest: harRequest1, responseBody, options });
  const leaf2 = createLeaf({ harRequest: harRequest2, responseBody, options });
  const merged = mergeLeaves(leaf1, leaf2);
  return merged;
};

it("merges leaves", () => {
  const options = defaultOptions;
  const result = setup(options);
  expect(result).toMatchSnapshot();
});

it("sets mostRecentRequest and mostRecentResponse", () => {
  const options = { ...defaultOptions, enableMoreInfo: true };
  const result = setup(options);
  expect(
    result.methods["POST"]?.["request"]!["application/json"]
      ?.mostRecent
  ).toEqual({ test: true });
  expect(
    result.methods["POST"]?.["response"]?.["200"]!["application/json"]
      ?.mostRecent
  ).toEqual({ test: 1 });
});
