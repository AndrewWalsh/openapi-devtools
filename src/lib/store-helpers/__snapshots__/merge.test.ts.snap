// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`merges leaves 1`] = `
{
  "methods": {
    "POST": {
      "queryParameters": {
        "properties": {
          "alt": {
            "type": "string",
          },
          "key": {
            "type": "string",
          },
        },
        "required": [
          "alt",
          "key",
        ],
        "type": "object",
      },
      "request": {
        "application/json": {
          "body": {
            "properties": {
              "test": {
                "type": "boolean",
              },
            },
            "required": [
              "test",
            ],
            "type": "object",
          },
        },
      },
      "requestHeaders": {
        "properties": {
          "anonymousid": {
            "type": "string",
          },
        },
        "required": [
          "anonymousid",
        ],
        "type": "object",
      },
      "response": {
        "200": {
          "application/json": {
            "body": {
              "properties": {
                "test": {
                  "type": "integer",
                },
              },
              "required": [
                "test",
              ],
              "type": "object",
            },
            "mostRecent": undefined,
          },
        },
      },
      "responseHeaders": undefined,
    },
  },
  "pathname": "/v1/track",
}
`;
